crates/bevy_ecs/src/storage/resource_non_send.rs

```diff
@@ -0,0 +1,595 @@
+            std::any::type_name::<R>()
+        );
+
+        OwningPtr::make(value, |ptr| {
+            // SAFETY: pointer is of type R
+            unsafe {
+                self.storage
+                    .get_mut(id)
+                    .map(|info| info.insert_with_ticks(ptr, ticks))
+                    .unwrap();
+            }
+        });
+
+        result
+    }
+
+    pub(crate) fn update_change_tick(&mut self) {
+        let tick = self.curr_tick.get();
+        self.curr_tick.set(tick.wrapping_add(1));
+    }
+}
+
+/// A "scoped lock" on [`ThreadLocals`], which is protected by a mutex and can be accessed
+/// through this guard via its [`Deref`] and [`DerefMut`] implementations.
+///
+/// When this guard is dropped, the lock will be unlocked.
+#[doc(hidden)]
+pub struct ThreadLocalsGuard<'a> {
+    guard: std::sync::MutexGuard<'a, ThreadLocals>,
+    // needed to decrement the strong reference count once dropped
+    _arc: Arc<Mutex<ThreadLocals>>,
+}
+
+impl std::ops::Deref for ThreadLocalsGuard<'_> {
+    type Target = ThreadLocals;
+    fn deref(&self) -> &Self::Target {
+        &*self.guard
+    }
+}
+
+impl std::ops::DerefMut for ThreadLocalsGuard<'_> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut *self.guard
+    }
+}
+
+/// Type alias for tasks that access thread-local data.
+pub type ThreadLocalTask = Box<dyn FnOnce(&mut ThreadLocals) + Send + 'static>;
+
+/// An error returned from the [`ThreadLocalTaskSender::send_task`] function.
+///
+/// A send operation can only fail if the receiving end of a channel is disconnected,
+/// implying that the data could never be received. The error contains the data that
+/// was sent as a payload so it can be recovered.
+#[derive(Debug)]
+pub struct ThreadLocalTaskSendError<T>(pub T);
+
+/// Channel for sending [`ThreadLocalTask`] instances.
+pub trait ThreadLocalTaskSender: DynClone + Send + 'static {
+    /// Attempts to send a task over this channel, returning it back if it could not be sent.
+    fn send_task(
+        &mut self,
+        task: ThreadLocalTask,
+    ) -> Result<(), ThreadLocalTaskSendError<ThreadLocalTask>>;
+}
+
+clone_trait_object!(ThreadLocalTaskSender);
+
+/// A [`Resource`] that enables the use of the [`ThreadLocal`] system parameter.
+#[derive(Resource)]
+struct ThreadLocalChannel {
+    owning_thread: std::thread::ThreadId,
+    direct: Weak<Mutex<ThreadLocals>>,
+    indirect: Box<dyn ThreadLocalTaskSender>,
+}
+
+// SAFETY: The pointer to the thread-local storage is only dereferenced in its owning thread.
+unsafe impl Send for ThreadLocalChannel {}
+
+// SAFETY: The pointer to the thread-local storage is only dereferenced in its owning thread.
+// Likewise, all operations require an exclusive reference, so there can be no races.
+unsafe impl Sync for ThreadLocalChannel {}
+
+/// Guards access to its thread-local data storage.
+pub struct ThreadLocalStorage {
```

Comment:
what's the advantage of the approach here vs building an abstraction around a global `thread_local!()`?